core/profiles/src/main/java/software/amazon/awssdk/profiles/internal/ProfileFileRefresher.java

```diff
@@ -0,0 +1,268 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.awssdk.profiles.internal;
+
+import java.io.IOException;
+import java.io.UncheckedIOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.time.Clock;
+import java.time.Instant;
+import java.util.Objects;
+import java.util.function.Consumer;
+import java.util.function.Function;
+import java.util.function.Supplier;
+import software.amazon.awssdk.annotations.SdkInternalApi;
+import software.amazon.awssdk.annotations.SdkTestInternalApi;
+import software.amazon.awssdk.profiles.ProfileFile;
+import software.amazon.awssdk.utils.SdkAutoCloseable;
+import software.amazon.awssdk.utils.cache.CachedSupplier;
+import software.amazon.awssdk.utils.cache.RefreshResult;
+
+/**
+ * Class used for caching and reloading ProfileFile objects from a Supplier.
+ */
+@SdkInternalApi
+public final class ProfileFileRefresher implements SdkAutoCloseable {
+
+    private final CachedSupplier<ProfileFileRefreshRecord> profileFileCache;
+    private volatile ProfileFileRefreshRecord currentRefreshRecord;
+    private final ProfileFileRefreshRecord emptyRefreshRecord;
+    private final Supplier<ProfileFile> profileFile;
+    private final Path profileFilePath;
+    private final Function<RuntimeException, ProfileFile> exceptionHandler;
+    private final Consumer<ProfileFile> onProfileFileReload;
+    private final Clock clock;
+
+    private ProfileFileRefresher(Builder builder) {
+        this.exceptionHandler = builder.exceptionHandler;
+        this.clock = builder.clock;
+        this.profileFile = builder.profileFile;
+        this.profileFilePath = builder.profileFilePath;
+        this.onProfileFileReload = builder.onProfileFileReload;
+        this.profileFileCache = CachedSupplier.builder(this::refreshResult)
+                                              .clock(this.clock)
+                                              .build();
+        this.emptyRefreshRecord = ProfileFileRefreshRecord.builder()
```

Comment:
nit: looks like this can be a static constant

Reply:
Good catch.